return unauthorized_client in oauth2 error code

strehle opened the pull request:
return unauthorized_client in oauth2 error response 

Fix for issue https://github.com/cloudfoundry/uaa/issues/2545

strehle committed:
return unauthorized_client in oauth2 error response code

Tallicia commented on uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/TokenMvcMockTests.java:
```diff
@@ -858,7 +858,7 @@ void token_endpoint_should_return_Basic_WWW_Authenticate_Header() throws Excepti
                         header()
                                 .stringValues("WWW-Authenticate",
-                                        "Basic realm=\"UAA/client\", error=\"unauthorized\", error_description=\"Bad credentials\"")
+                                        "Basic realm=\"UAA/client\"")
```
Is this meant to remove? Instead of replace or adjust the error description?

strehle replied:
solved

Tallicia reviewed (changes requested):
I want to validate the one question inline before approving.

strehle committed:
Ensure www-authenticate header

In case of an exception there is a generic
www-authenticate header which ends in a browser
popup. For Oauth2 errors this is normally not
required. Ensure with this commit to stay
with same header message

strehle commented:
@bruce-ricard since you are active on some issues, can you please check if you see an issue with this, I would like to solve this for next release

neelalex commented:
@strehle Should this be `invalid_client` as you mentioned in the ticket?

strehle commented:
> @strehle Should this be `invalid_client` as you mentioned in the ticket?

not necessarily, but currently used unauthorized_client to have it similar to non stand unauthorized code.

My tests with others have shown, that they use invalid_client, but unauthorized_client  is also oauth2 complaint.

But therefore asked the other UAA maintainer for their thouths

neelalex commented:
> > @strehle Should this be `invalid_client` as you mentioned in the ticket?
> 
> not necessarily, but currently used unauthorized_client to have it similar to non stand unauthorized code.
> 
> My tests with others have shown, that they use invalid_client, but unauthorized_client is also oauth2 complaint.
> 
> But therefore asked the other UAA maintainer for their thouths

The issue addressed in this ticket is **with invalid client credentials**, the error response was expected to be as follows as per https://www.rfc-editor.org/rfc/rfc6749#section-5.2

Code: 401 Unauthorized
Body:
**“error”: “invalid_client”**
“error_description”: “Bad credentials”


whereas ->  unauthorized_client : The authenticated client is not authorized to use this authorization grant type.
This means  this error occurs when the client is **not authorized to request the requested resource, even if the client credentials are valid**. It typically indicates that the client does not have the necessary permissions or scope to access the requested resource.

Tallicia commented:
Thank you for clarifying this @neelalex

strehle commented:
close this, because we have clarifid we go with the alternative PR https://github.com/cloudfoundry/uaa/pull/2596
